src/control/lib/daos/api/libdaos_selftest_stubs.go

```diff
@@ -0,0 +1,170 @@
+//
+// (C) Copyright 2024 Intel Corporation.
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+//
+//go:build test_stubs
+// +build test_stubs
+
+package api
+
+import (
+	"unsafe"
+
+	"github.com/daos-stack/daos/src/control/lib/daos"
+	"github.com/daos-stack/daos/src/control/lib/hostlist"
+	"github.com/daos-stack/daos/src/control/lib/ranklist"
+)
+
+/*
+#cgo CFLAGS: -I${SRCDIR}/../../../../cart/utils -I${SRCDIR}/../../../../utils/self_test
+#cgo LDFLAGS: -lgurt -lcart -ldaos_self_test
+
+#include <daos_errno.h>
+#include <gurt/common.h>
+
+#include "self_test_lib.h"
+
+struct st_latency ***
+alloc_latency_arrays(size_t num_sizes, size_t num_endpoints, size_t num_latencies)
+{
+	struct st_latency	***latencies = NULL;
+	int					   i, j;
+
+	D_ALLOC_ARRAY(latencies, num_sizes);
+	if (latencies == NULL)
+		return NULL;
+
+	for (i = 0; i < num_sizes; i++) {
+		D_ALLOC_ARRAY(latencies[i], num_endpoints);
+		if (latencies[i] == NULL)
+			return NULL;
+
+		for (j = 0; j < num_endpoints; j++) {
+			D_ALLOC_ARRAY(latencies[i][j], num_latencies);
+			if (latencies[i][j] == NULL)
+				return NULL;
+		}
+	}
+
+	return latencies;
+}
+*/
+import "C"
+
+type run_self_test_EndpointLatency struct {
+	val    C.int64_t
+	rank   C.uint32_t
+	tag    C.uint32_t
+	cci_rc C.int
+}
+
+var (
+	run_self_test_RunConfig         *daos.SelfTestConfig
+	run_self_test_RC                C.int = 0
+	run_self_test_MsEndpoints       []daos.SelfTestEndpoint
+	run_self_test_EndpointLatencies []run_self_test_EndpointLatency
+)
+
+func run_self_test(sizes *C.struct_st_size_params, numSizes C.int, repCount C.int, maxInflight C.int, groupName *C.char,
```

Comment:
As I'm looking at the complicated logic in some of these stubs to convert between C and Go, I wonder if it wouldn't be better to write the stubs in C, and just call them from the Go stubs.

Reply:
For some of the stubs to come as more of the API gets added to these bindings, I definitely agree. For this one in particular, at least some of the logic still has to be in Go in order to reconstruct the test configuration from the function arguments. It was also just faster for me personally to do it all in Go. :)